### PR TITLE
[SSHD-721] I/O workers exhaustion in tcpip forwarding

### PR DESCRIPTION
**Pull request by gnodet**

### CODE REVIEW

File: sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Port forwarding test multiple clients connecting at once.
+ */
+public class ConcurrentConnectionTest extends BaseTestSupport {
```
Comment (lgoldstein) [inline sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java]: Missing @FixMethodOrder annotation

File: sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    private static final byte[] PAYLOAD_TO_SERVER = "To Server -> To Server -> To Server".getBytes();
+    private static final byte[] PAYLOAD_TO_CLIENT = "<- To Client <- To Client <-".getBytes();
+    private static final Logger LOG = LoggerFactory.getLogger(ConcurrentConnectionTest.class);
```
Comment (lgoldstein) [inline sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java]: Let's use `getBytes(StandardCharsets.UTF_8)` so we are aligned with the rest of the code

File: sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+    // Test Server State
+    private int testServerPort;
+    private ServerSocket testServerSock;
+    private Thread testServerThread;
+
+    // SSH Client State
+    private ClientSession session;
+
+    /*
```
Comment (lgoldstein) [inline sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java]: Isn't _checkstyle_ requiring a public default no-args constructor ?

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java [hunk 4]
```diff
@@ -979,15 +980,22 @@ public void messageReceived(IoSession session, Readable message) throws Exceptio
...
+                outputStream.flush();
+            } else {
+                future.addListener(f -> {
+                    try {
+                        OutputStream outputStream = channel.getInvertedIn();
+                        outputStream.write(buffer.array(), buffer.rpos(), buffer.available());
+                        outputStream.flush();
+                    } catch (IOException e) {
+                        channel.getSession().exceptionCaught(e);
+                    }
```
Comment (lgoldstein) [inline sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java]: I prefer having separate statements for ease of debugging:
```java
} catch (IOException e) {
    Session s = channel.getSession();
    s.exceptionCaught(e);
}
```

File: sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+    protected void serverAcceptLoop() {
+        try {
+            final AtomicInteger activeServers = new AtomicInteger(0);
+            testServerSock = new ServerSocket(0);
+            testServerPort = testServerSock.getLocalPort();
+            LOG.debug("Listening on {}", testServerPort);
+            while (true) {
+                final Socket s = testServerSock.accept();
+                LOG.debug("Got connection");
```
Comment (lgoldstein) [inline sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java]: No need for _final_

File: sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+                server.setDaemon(true);
+                server.setName("Server " + s.getPort());
+                server.start();
+            }
+        } catch (final SocketException e) {
+            LOG.debug("Shutting down test server");
+        } catch (final Throwable t) {
+            LOG.error("Error", t);
+        }
+    }
```
Comment (lgoldstein) [inline sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java]: No need for _final_

File: sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+            final long r = s.getInputStream().read(buf);
+            LOG.debug("Read {} payload from client", r);
+            s.getOutputStream().write(PAYLOAD_TO_CLIENT);
+            LOG.debug("Wrote payload to client");
+            s.close();
+            LOG.debug("Active Servers: {}", activeServers.decrementAndGet());
+        } catch (final Throwable t) {
+            LOG.error("Error", t);
+        }
+    }
```
Comment (lgoldstein) [inline sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java]: No need for _final_

File: sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    public static void stopServer() throws IOException {
+        if (!server.close(true).await(TIMEOUT)) {
+            LOG.warn("Failed to close server within {} sec.", TimeUnit.MILLISECONDS.toSeconds(TIMEOUT));
+        }
+    }
+
+    @Before
+    public void createClient() throws IOException {
+        final SshClient client = SshClient.setUpDefaultClient();
+        client.setForwardingFilter(AcceptAllForwardingFilter.INSTANCE);
```
Comment (lgoldstein) [inline sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java]: No need for _final_

File: sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+    @Test
+    /*
+     * Run PORT_FORWARD_CLIENT_COUNT simultaneous server threads.
+     * 
+     * Emulates a web browser making a number of simultaneous requests on
+     * different connections to the same server HTTP specifies no more than two,
+     * but most modern browsers do 6 or more.
+     */
+    public void testConcurrentConnectionsToPortForward() throws Exception {
+        final SshdSocketAddress remote = new SshdSocketAddress(TEST_LOCALHOST, 0);
```
Comment (lgoldstein) [inline sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java]: I think it is easier to read if documentation precedes the annotation (i.e., annotation is attached to the method). BTW, is _checkstyle_ allowing this ?

File: sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+            final Thread t = new Thread(() -> {
+                try {
+                    bytesRead[n] = makeClientRequest(forwardedPort, b, wait);
+                    LOG.debug("Complete, received full payload from server.");
+                    success.incrementAndGet();
+                } catch (final Exception e) {
+                    fail.incrementAndGet();
+                    LOG.error("Error in client code", e);
+                }
+            });
```
Comment (lgoldstein) [inline sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java]: No need for _final_

File: sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        final long r = s.getInputStream().read(buf);
+        LOG.debug("Read {} payload from server", r);
+
+        assertEquals("Mismatched data length", PAYLOAD_TO_CLIENT.length, r);
+        s.close();
+
+        return r;
+    }
+
+}
```
Comment (lgoldstein) [inline sshd-core/src/test/java/org/apache/sshd/common/forward/ConcurrentConnectionTest.java]: Superfluous empty line

**Comment by gnodet** [comment]
Feel free to modify the code to suit your taste.  The unit test works as designed, the build passes all the checkstyle rules, it compiles the test succeeds :-)  And it has been merged already...

Fwiw, what I'd like is to set up PR automatic tests, I need to check how to do that at the ASF.

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java (hunk 4 above)
Comment (bkuker) [reply sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java]: I am getting suspicious of this line....

Taking my (web browser -> ssh server <-> ssh client -> web server) example, when I kill -9 my **web browser** while it has open connections through the port forward, this line seems to be causing the ssh server to drop the connection to the ssh client....

**Pull request merged**